modules/internal/localeModels/apu/LocaleModel.chpl

```diff
@@ -0,0 +1,295 @@
+      f <~> '.'+name;
+    }
+
+    proc getChildCount(): int { return 0; }
+    iter getChildIndices() : int {
+      halt("No children to iterate over.");
+      yield -1;
+    }
+    proc addChild(loc:locale) {
+      halt("Cannot add children to this locale type.");
+    }
+    proc getChild(idx:int) : locale { return nil; }
+  }
+
+  class GPULocale : AbstractLocaleModel {
+    const sid: chpl_sublocID_t;
+    const name: string;
+
+    proc chpl_id() return (parent:LocaleModel)._node_id; // top-level node id
+    proc chpl_localeid() {
+      return chpl_buildLocaleID((parent:LocaleModel)._node_id:chpl_nodeID_t,
+                                sid);
+    }
+    proc chpl_name() return name;
+
+    proc GPULocale() {
+    }
+    proc ~GPULocale() {
+     extern proc hsa_shutdown(): void ;
+
+     // Comment out this until runtime support for HSA is added
+
+     //     hsa_shutdown();
+    }
+
+
+    proc GPULocale(_sid, _parent) {
+      sid = _sid;
+      parent = _parent;
+      name = parent.chpl_name() + "-GPU" + sid;
+    }
+
+    proc writeThis(f) {
+      parent.writeThis(f);
+      f <~> '.'+name;
+    }
+
+    proc getChildCount(): int { return 0; }
+    iter getChildIndices() : int {
+      halt("No children to iterate over.");
+      yield -1;
+    }
+    proc addChild(loc:locale) {
+      halt("Cannot add children to this locale type.");
+    }
+    proc getChild(idx:int) : locale { return nil; }
+  }
+
+  const chpl_emptyLocaleSpace: domain(1) = {1..0};
+  const chpl_emptyLocales: [chpl_emptyLocaleSpace] locale;
+
+  //
+  // The node model
+  //
+  class LocaleModel : AbstractLocaleModel {
+    const _node_id : int;
+    const local_name : string;
+
+    const numSublocales: int;
+    var GPU: GPULocale;
+    var CPU: CPULocale;
+
+    // This constructor must be invoked "on" the node
+    // that it is intended to represent.  This trick is used
+    // to establish the equivalence the "locale" field of the locale object
+    // and the node ID portion of any wide pointer referring to it.
+    proc LocaleModel() {
+      if doneCreatingLocales {
+        halt("Cannot create additional LocaleModel instances");
+      }
+      setup();
+    }
+
+    proc LocaleModel(parent_loc : locale) {
+      if doneCreatingLocales {
+        halt("Cannot create additional LocaleModel instances");
+      }
+      parent = parent_loc;
+      setup();
+    }
+
+    proc chpl_id() return _node_id;     // top-level locale (node) number
+    proc chpl_localeid() {
+      return chpl_buildLocaleID(_node_id:chpl_nodeID_t, c_sublocid_any);
+    }
+    proc chpl_name() return local_name;
+
+
+    proc writeThis(f) {
+      // Most classes will define it like this:
+      //      f <~> name;
+      // but here it is defined thus for backward compatibility.
+      f <~> new ioLiteral("LOCALE") <~> _node_id;
+    }
+
+    proc getChildSpace() {
+      halt("error!");
+      return {0..#numSublocales};
+    }
+
+    proc getChildCount() return 0;
```

Comment:
It's clear that these getChild functions are inconsistent. @dmk42 - we had some plan to allow for querying sublocales by type. Do you have written down somewhere what that interface will look like? We should at least start a github issue about it.

Anyway, right now code like the range/default rectangular array parallel iterators will use getChildCount() to compute the number of sublocales (https://github.com/chapel-lang/chapel/blob/master/modules/internal/ChapelRange.chpl#L1713). We don't want it to by default share work between the CPU and the GPU (and we want it to only run on the GPU when requested), so I'd recommend treating the GPU similarly to the memory types in the KNL locale model. That is, I don't think the GPU sublocale should be a "child" of this locale... (at least not until we fix up the interface).

Clearly we have some interface work to do. I don't think this issue has to block merging the APU locale model, but I'd like too see it addressed. At the very least, it needs to be covered in an issue.

@dmk42 - what do you think about this?